### ISSUE TITLE
Something wrong with png

### ISSUE THREAD

**Issue by codeservice** [haxenme/nme#508]
It was ok, on older nme version 
![screenshot - 03_10_2017 11_29_39 am](https://user-images.githubusercontent.com/1771547/31133740-51385cbe-a82e-11e7-99c4-1c144cf4f25f.jpg)

![image1](https://user-images.githubusercontent.com/1771547/31133885-aee53a12-a82e-11e7-8ddb-57c17f17a49d.png)

**Comment by madrazo** [comment]
Is it a 32 bit (with alpha) or without alpha? Is it the Windows (angle) target or other?

**Comment by codeservice** [comment]
png attached (second image) first image screenshot
compiled on windows 7 x86

I am testing nme samples "03-Bitmaps" all images here

**Comment by madrazo** [comment]
OK. Not so long I worked around with this https://github.com/haxenme/nme/pull/483/files
but I though it was fixed now  (dont' forget to recompile ndll if you try it project>neko build.n ndll-windows) :D

**Comment by codeservice** [comment]
I dont use ANGLE, I compiled libraries without it

**Comment by madrazo** [comment]
Sorry, how can I run the "03-Bitmaps" sample? I'm getting error "This sample shows how to use resources without the nme tool."

**Comment by codeservice** [comment]
I created nmml file and changed code a bit.
I think easiest way simple load this image in any new test project with:

      var bytes = nme.utils.ByteArray.readFile(inFilename);
      return BitmapData.loadFromBytes(bytes);

I dont use Angle for now because its not available for all platforms yet. I need cross platform consistency.

**Comment by hughsando** [comment]
Do you have this fix:
https://github.com/haxenme/nme/commit/60ecdabd3eabfec2479af6221498670b00f67cb1#diff-7d05115c02f5f6b21040987b75abe209
?

**Comment by codeservice** [comment]
Yes, all lines from this fix in place. I dont see issues with jpg, only png.
png toolkit updated from git, no differences here too. I have some custom
modifications in nme. All not related to graphics rendering. Old nme with
the same customizations works fine.

I updated sdl to 2.0.6 maybe problem here? My old nme version works well
with 2.0.6

**Comment by hughsando** [comment]
No, this is certainly an internal nme pixel format issue.
It looks like a RGBA is being interpreted as RGB or the RGBA is being
interpreted as a RGB.
Setting the format explicitly may work-around this issue.

bitmapData.format = nme.image.PixelFormat.pfBGRA;

**Comment by codeservice** [comment]
Using bitmapData.format = nme.image.PixelFormat.pfBGRA; fixing problem.
Its strange issue. format not recognized in some conditions. If I do new
clean sample it works well. In original sample its broken. Trying to find
out what exactly brakes format detection.

Clean simple example works:

var bytes = nme.utils.ByteArray.readFile("C:\\Temp\\Image1.png");
var bd: nme.display.BitmapData =
nme.display.BitmapData.loadFromBytes(bytes, null);
//bd.format = nme.image.PixelFormat.pfBGRA;
var bm: nme.display.Bitmap = new nme.display.Bitmap(bd);
Lib.current.stage.addChild(bm);

**Comment by hughsando** [comment]
I added some pre-processing to png images, so they now have an
nme.AlphaMode.
The idea is that the tool can convert RGBA to premultiplied to speed load
times and make the files smaller, or tell the asset to premultiply on load.
Now, if the original png was RGB, not RGBA then some case may not have been
done correctly.

But the fact that now projects work suggests to me that there is an old
file in the bin directory, but the tool has provided a new alpha mode and
when it does:
data.setFormat( nme.image.PixelFormat.pfBGRPremA, false );
May be forcing an rgba format on an RGB image.
Does clearing the bin directory fix this?

Hugh

**Comment by codeservice** [comment]
I'll try clean bin directory. I was able replicate it in different project copying code from sample.
This issue repeatable only when I have this png image with other graphics on screen. I had FPS text info. This info has two text fields, so if I left one text field on screen png works, two or more - broken. Really strange.
I have modifications in text, maybe I did something wrong.

**Comment by hughsando** [comment]
That sounds like some format attribute is not getting cleared or
initialized in the the opengl driver code.

**Comment by codeservice** [comment]
If I remove text fields from screen in hx code and next line add png it will be broken. It's like text fields changing gl rendering once they both on screen and this change stays active all the time even if I remove all text from stage. 
Any hint what else to check?

**Comment by hughsando** [comment]
It is a bit of a strange one without reproduction case.
The stateful error suggests some uninitialized thing, and the most stateful code is in OGLData::RenderData.
Here it tries to cache pixel programs and buffer data, and does some stuff with BlendFunc which could be wrong.
But, the images you show are to do with BitmapData pixel format, which is not very stateful as far as I can tell.  All I can think of is that the gl texture Id is somehow crossing over with the font and influencing the gl format - so some call to "texture->Bind" might be missing somewhere.
If you can send me a sample I can have a go at debugging.
Otherwise, I think printfing the OGLTexture fromats
GLuint pixel_format = getTransferOgl(fmt);
and mUploadedFormat, and see if they differ between working and non-working and trace backwards from there.

**Comment by codeservice** [comment]
I dont know how to replicate on your end but I know how to fix it. 
I compared old and new OGLTexture.cpp and found one big difference how texture initialized.
In method "Bind" you have new line:
glPixelStorei(GL_UNPACK_ALIGNMENT, 1);

You set unpack alignment to 1 which is not default value (4)
If I simply remove this line or use value 4 everything works correctly.

I think right solution restore default value 4 when texture without pixels binding:

            else
            {
               #ifndef NME_GLES
               glPixelStorei(GL_UNPACK_ROW_LENGTH, mSurface->Width());
               glPixelStorei(GL_UNPACK_ALIGNMENT, 1); //PROBLEM Spot !!!!
               #endif
               glTexSubImage2D(GL_TEXTURE_2D, 0,
                  x0, y0,
                  dw, dh,
                  pixel_format, channel,
                  mSurface->Row(y0) + x0*pw );
               #ifndef NME_GLES
               glPixelStorei(GL_UNPACK_ROW_LENGTH,0);
               #endif
            }

            int err = glGetError();
            if (err != GL_NO_ERROR)
               ELOG("GL Error: %d %dx%d", err, mDirtyRect.w, mDirtyRect.h);
            mDirtyRect = Rect();
         }
      }
      else  {glPixelStorei(GL_UNPACK_ALIGNMENT, 4); //FIX here !!!!!!!!!!!!!!!!!!!!!!!!
         glBindTexture(GL_TEXTURE_2D,mTextureID);
         }
